### DISCUSSION TITLE
Compile enviroments variables into nuxt at build time

### DISCUSSION THREAD

**Discussion by aeropagz** [nuxt/nuxt#18501] [Questions]
Hello,
I want to compile env vars into my code at build time. These are values like the git hash, build-time etc and are available in a pipeline that builds an docker image. 

I am aware of runtimeConfig but thats not helping because at runtime the env vars are not set anymore

**Comment by dsvgl** [comment]
Have you read https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables?

If I understand you correctly, it could be done this way:

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  runtimeConfig: {
    public: {
      gitHash: process.env.GIT_HASH || '',
    }
  },
})
```

```vue
<!-- some Vue component -->
<template>
  <div>Git hash is: {{ gitHash }}</div>
</template>

<script setup lang="ts">
const { gitHash } = useRuntimeConfig().public;
</script>
```

**Reply by aeropagz** [reply]
Yes I've seen that. But the problem is that you can change the environment variable now at runtime what I don't want. It should be compiled with webpack/vite only at build time.

**Reply by dsvgl** [reply]
Ah yeah. appConfig is bundled, but doesn't support reading env vars :/
https://nuxt.com/docs/getting-started/configuration#runtimeconfig-vs-appconfig

If you have env vars starting with `VITE_` then they are accessible in your app and are burned in at build-time.

**Reply by aeropagz** [reply]
Yes App config would be great if env are supported.

I think I have now a solution. I use Vite's define in nuxt config.

There you can store env vars in meta.env.
It's not very clean but it should work.

**Reply by dsvgl** [reply]
Could you show a snippet? I also tried defining vars via `define` but it didn't worked. thx

**Reply by aeropagz** [reply]
yes sure: 
```ts
export default defineNuxtConfig({
  // ....
  vite: {
    define: {
      "import.meta.env.PACKAGE_VERSION": JSON.stringify(process.env.npm_package_version),
      "import.meta.env.CI_PROJECT_NAME": JSON.stringify(process.env.CI_PROJECT_NAME),
      "import.meta.env.CI_COMMIT_REF_NAME": JSON.stringify(process.env.CI_COMMIT_REF_NAME),
      "import.meta.env.CI_COMMIT_SHORT_SHA": JSON.stringify(process.env.CI_COMMIT_SHORT_SHA),
      "import.meta.env.CI_JOB_STARTED_AT": JSON.stringify(process.env.CI_JOB_STARTED_AT),
      "import.meta.env.CI_COMMIT_TIMESTAMP": JSON.stringify(process.env.CI_COMMIT_TIMESTAMP)
    }
  },
});
```

**Reply by dsvgl** [reply]
thx!

**Reply by gabrielmoris** [reply]
Hello, I have the same Problem and I would like to try your solution. In the app : How could I access to this env variables?

**Reply by aeropagz** [reply]
you would access like this:
```js
const packageVersion = import.meta.env.PACKAGE_VERSION;
```
[See](https://vitejs.dev/guide/env-and-mode.html) for further info

**Reply by gabrielmoris** [reply]
Thank you!

**Reply by unitythemaker** [reply]
Is there a way to make ``vite.define`` work on server-side codes?
